pkg/cmd/validate.go

```diff
@@ -140,69 +176,79 @@ func (c *commandFlags) Run(cmd *cobra.Command, args []string) error {
 	if c.outputFormat == OutputHuman {
 		for _, path := range files {
 			fmt.Fprintf(cmd.OutOrStdout(), "\n\033[1m%v\033[0m...", path)
-			e := ValidateFile(path, factory)
-			if e != nil {
+			var errs []error
+			for _, err := range ValidateFile(path, factory) {
+				if err != nil {
+					errs = append(errs, err)
+				}
+			}
+			if len(errs) != 0 {
 				fmt.Fprintln(cmd.OutOrStdout(), "\033[31mERROR\033[0m")
-				fmt.Fprintln(cmd.ErrOrStderr(), e.Error())
+				for _, err := range errs {
+					fmt.Fprintln(cmd.ErrOrStderr(), err.Error())
+				}
 			} else {
 				fmt.Fprintln(cmd.OutOrStdout(), "\033[32mOK\033[0m")
 			}
 		}
 	} else {
-		res := map[string]metav1.Status{}
+		res := map[string][]metav1.Status{}
 		for _, path := range files {
-			valErr := ValidateFile(path, factory)
-			var statusErr *k8serrors.StatusError
-			var fieldError *field.Error
-			var errorList utilerrors.Aggregate
-
-			if errors.As(valErr, &statusErr) {
-				res[path] = statusErr.ErrStatus
-			} else if errors.As(valErr, &fieldError) {
-				res[path] = k8serrors.NewInvalid(schema.GroupKind{}, "", field.ErrorList{fieldError}).ErrStatus
-			} else if errors.As(valErr, &errorList) {
-				errs := errorList.Errors()
-				var fieldErrs []*field.Error
-				var otherErrs []error
-				for _, e := range errs {
-					fieldError = nil
-					if errors.As(e, &fieldError) {
-						fieldErrs = append(fieldErrs, fieldError)
-					} else {
-						otherErrs = append(otherErrs, e)
-					}
-				}
-				if len(otherErrs) == 0 {
-					res[path] = k8serrors.NewInvalid(schema.GroupKind{}, "", fieldErrs).ErrStatus
-				} else {
-					res[path] = k8serrors.NewInternalError(errors.Join(otherErrs...)).ErrStatus
-				}
-
-			} else if valErr != nil {
-				res[path] = k8serrors.NewInternalError(valErr).ErrStatus
-			} else {
-				res[path] = metav1.Status{Status: metav1.StatusSuccess}
+			for _, err := range ValidateFile(path, factory) {
+				res[path] = append(res[path], errorToStatus(err))
 			}
 		}
 		data, e := json.MarshalIndent(res, "", "    ")
 		if e != nil {
 			return fmt.Errorf("failed to render results into JSON: %w", e)
 		}
-
 		fmt.Fprintln(cmd.OutOrStdout(), string(data))
 	}
 
 	return nil
 }
 
-func ValidateFile(filePath string, resolver *validatorfactory.ValidatorFactory) error {
+func ValidateFile(filePath string, resolver *validatorfactory.ValidatorFactory) []error {
```

Comment:
@alexzielenski is this a good approach to return a slice of errors ?

Reply:
Yes but I think the skipped documents should leave `nil` error element so the returned errors matches and ordering and amount of input documents

Reply:
But the `nil` error will result in a success `Status` entry, looks strange as it didn't exercise the validation process at all (a skipped document is one that contains only comments right ?)

Reply:
That is true. I am concerned about the confusing interface where input documents != output statuses. What do you think of this idea:

Rather than return a slice of errors, return a map of GVKNN (group-verison-kind-namespace-name) to error. This would make it very clear which error is associated with which document. Skipped documents would not need to be included. The human output would remain the same, but the JSON output contains clear machine-readable association between document and error.

Reply:
I thought about that but it is allowed to have multiple identical GVKNN documents in the same file.

Reply:
Rats. Tough. I am still learning towards recording empty documents as success. It looks like for the human output we only log the non-nil errors so for people using the tool directly it's not that confusing. For JSON output I think it makes most sense to keep `Success` status for empty docs.